File: Data Structures & Algorithms/balanced-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        return this.dfs(root)[0]


    }

    dfs(root) {
        if (!root) return [true, 0];
        let [left, right] = [this.dfs(root.left), this.dfs(root.right)]
        const balanced = (Math.abs(left[1] - right[1]) <= 1) && left[0] && right[0];
        return [balanced, 1 + Math.max(left[1], right[1])];
    }

}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = float('-inf')
        def dfs(root):
            nonlocal ans
            if not root:
                return float('-inf')
            left_max_path_sum = dfs(root.left)
            right_max_path_sum = dfs(root.right)
            ans = max(ans, left_max_path_sum, right_max_path_sum, root.val + left_max_path_sum,root.val + right_max_path_sum, left_max_path_sum + right_max_path_sum + root.val, root.val)
            return max(0, left_max_path_sum, right_max_path_sum) + root.val
        dfs(root)
        return ans